### ISSUE TITLE
Disallow backslash as an operator

### ISSUE THREAD

**Issue by LiamGoodacre** [purescript/purescript#3517]
This would let us commit earlier to parsing something as a lambda abstraction and potentially improve parse errors/performance.  Backslash could still be used in a larger operator name, but just not by itself.  So `\` isn't an operator, but `\\` could be.

**Comment by natefaubion** [comment]
I'm co-opting this threat to list potential deprecations for all the dark corner-cases in the current grammar.

**Comment by natefaubion** [comment]
We should disallow `@` as an operator. It conflicts with binder syntax.

**Comment by natefaubion** [comment]
Somewhat contentious I'm sure, but I think we should disallow any whitespace besides `\n`, `\r\n`, and ASCII space (except in raw string literals). Unicode whitespace and tabs are problematic for layout, and I don't think there's a reason to support it.

**Comment by natefaubion** [comment]
We should fix precedence of kind annotated types. Currently this is allowed and I don't even...
```purescript
type Foo a b c = a :: Type -> Type -> Type b :: Type c :: Type
```

**Comment by hdgarrood** [comment]
I suspect the single-branch case pattern is common in core libs because you didn’t used to be able to pattern match in a let binding, ie
```
let Foo bar = a
 in bar + 2
```
didn’t used to be legal, but that’s probably what I’d recommend now.

**Comment by garyb** [comment]
@hdgarrood yeah, that's exactly it.

**Comment by natefaubion** [comment]
We should disallow nested backtick expressions. This is currently allowed:
```purescript
a `b `c` d` e
```
But this is ambiguous, since it can also be parsed as:
```purescript
(a `b` c) `d` e
```
Which I think is probably what one would want anyway.

**Comment by natefaubion** [comment]
We should disallow `forall` as a valid identifier. It's disallowed in types, so this means that type identifiers and value identifiers have different rules.

**Comment by natefaubion** [comment]
We should simplify string and character escapes. We currently use essentially what Haskell has (really it's whatever is provided by parsec in the default language parser implementation), but it has a lot of archaic escape codes and rules. We should only allow
* `\r`
* `\n`
* `\t`
* `\\`
* `\'`
* `\"`
* `\x[0-9a-f]{1,4}` (since we only allow utf-16 codepoints in char literals anyway)
* `\[\r\n ]+\` (whitespace gap)

**Comment by hdgarrood** [comment]
I’m not sure about restricting \x character escapes to 4 hex characters actually: that leaves you with no comfortable way of inserting astral plane characters (ie those with code point values greater than 0xffff) other than inserting the literal character. We should continue to disallow them in Char literals, since Chars are single code units rather than code points, but we should be able to treat strings as sequences of code points where appropriate, so I think “\x12345” should be valid (and equivalent to “\u{12345}” in JS).

I’d like to also copy JS’ thing where you can use braces to delimit the hex literal from subsequent characters for clarity.

**Comment by natefaubion** [comment]
FWIW, I think you could still use surrogate pairs. Perhaps then we should allow `\x` with 4 hex characters, and `\u{..}` syntax for arbitrary escape. The problem with not having a limit on `\x` is you have to have some way to decide when to stop considering something as part of the escape. Haskell has `\&` as a zero-width escape for this purpose.

**Comment by hdgarrood** [comment]
Yeah I agree that we would need to have something to indicate the end of the escape, but I don’t think we should need to change the syntax in a breaking way to achieve that. My preference would be to leave \x as it is currently (which is still useful even in the absence of \&, in the case where the escape is the last thing in the string, or the subsequent character isn’t in [0-9a-f]) but additionally add braces to allow sensible delimiting going forward.

**Comment by natefaubion** [comment]
Do we want to take this opportunity to get rid of `#` in kind syntax. It's somewhat problematic that kinds and types have different syntax, and I think unifying them might do some good.

**Comment by garyb** [comment]
It was left that way intentionally when the other kind symbols were removed because it's unique in being the kind that takes parameters, but I have nothing against naming it.

**Comment by natefaubion** [comment]
> It was left that way intentionally when the other kind symbols were removed because it's unique in being the only kind that takes parameters, but I have nothing against naming it.

Yes, that's true, but I think if we eventually want polykinds, it's a good idea to get rid of it. Or at least I don't think it adds anything. It's already "special" in that the language does not let you define a kind that takes a parameter.

**Comment by hdgarrood** [comment]
Would we need to implement kind polymorphism to be able to do that sensibly though? - if the row kind constructor `#` were able to exist on its own, what would its kind be? Also I’d want to deprecate it for a release series, not outright remove it with little warning.

**Comment by hdgarrood** [comment]
Also, rows are always going to be special, even if we get kind polymorphism and the ability to parameterize user-defined kinds, because they have built in support in the type checker, right? So I think it’s probably appropriate that the syntax reflects that.

**Comment by hdgarrood** [comment]
Wait, I’m talking nonsense, aren’t I? - kinds don’t themselves have kinds.
